keps/sig-api-machinery/2211-omit-managed-fields-server-side/README.md

```diff
@@ -0,0 +1,372 @@
+In addition, the bandwidth usage and size of internal caches could be significantly affected due to the additional managedFields.
+
+This KEP is proposing omitting the managedFields on the server-side when clients request objects to make the response less verbose
+and save the amount of data that needs to be transferred.
+
+## Motivation
+
+Many users have complained in [kubernetes#90066][90066] that they got impacted by the addition of managedFields
+(and the issue itself has received more than 100 thumb-up reactions), and we want to improve the user experience.
+
+[90066]: https://github.com/kubernetes/kubernetes/issues/90066
+
+At the same time, it is not ideal to only drop the managedFields on the client-side, because that requires the same
+bandwidth usage compared to simply returning the original object, and every client
+(such as kubectl or CLI tools written by users using client-go) needs to drop the managedFields by themselves.
+
+With this KEP, the API server could return objects without managedFields on demand, so that least client-side changes is needed
+to make the response less verbose. 
+
+### Goals
+
+* Omit managedFields from responses, when requested by clients.
+
+### Non-Goals
+
+* Turn off managedFields. Server-side Apply relies on managedFields, we could not simply disable it.
+* Clear manageFields only on the client-side.
+
+## Proposal
+
+### Server-side changes
+
+* Add three content types:
+  * `application/json+managed.omit`
+  * `application/yaml+managed.omit`
+  * `application/vnd.kubernetes.protobuf+managed.omit`
+* Add a new serializer, say `UnmanagedSerializer`, which wraps existing serializers and omits the managedFields before encoding objects.
+* Make the API server to use `UnmanagedSerializer` when clients properly set the `Accept` header and the selected
+  content type after negotiation is one of the above three.
+
+In addition, it is worth noting that we could add similar content types to represent specific format version of managedFields,
+such as `application/json+managed.v1`, `application/yaml+managed.v2` etc, so that users are able to request a certain
+version of managedFields with these headers.
+
+### Client-side changes
+
+Implement a flag in `kubectl` to let users choose to omit the managedFields.
+
+### Risks and Mitigations
+
+Risk:
+* The API server might not be up-to-date and cannot recognize the new content types.
+
+Mitigations:
+1. We could add the original content type to the `Accept` header as a fallback, such as
+  `Accept: application/json+managed.omit, application/json;q=0.9`, therefore the API server could return JSON-formatted 
+  objects rather than reject the request with a 406(Not Acceptable) error if it is too old to recognize the new content type.
+2. Then make `kubectl get` to drop managedFields if user wants to omit managedFields, but the API server doesn't realize that.
+
+## Design Details
+
+### Server-side
```

Comment:
it's not clear to me why the server-side piece is required... that seems significantly more complex and doesn't remove the need for client-side logic to pass alternate content-type headers. Is the bandwidth savings significant? Why would we not just drop the field in the client?

Reply:
Correct, it's ~30% of the object size if you're in proto mode.

Reply:
Now that you mention it though, why don't we just implement in kubectl and see if controller operators complain? So far all the complaints have just been from users that don't want to see it AFAIK.

Reply:
> why don't we just implement in kubectl and see if controller operators complain?

I would definitely start there. If there's good evidence we need to push it further up the stack, doing it with a `;omit=metadata.managedFields` modifier or something in a later iteration could be reasonable

Reply:
We already have complaints about the network impact: 
https://github.com/kubernetes/kubernetes/issues/90066#issuecomment-731452459
> Hi, @apelisse, Due to addition managedFields metadata, payload of the every object got increased by 40%. For example, for simple nginx pod, before <1.18, response was ~6KB and >=1.18, response is 10KB. Ours is monitoring application, we use k8s apis to get the response of all objects (chunked fashion) and we dont use this field. The additional payload significantly costing the cpu and memory. is there any way to express the k8s API to exclude in the response ? If not, having functionality like this will significantly helpful for the clients.

https://github.com/kubernetes/kubernetes/issues/90066#issuecomment-734142491
> As I've mentioned before and as @ganga1980 mentioned as well, the main managed fields problem is not a cosmetic one of hiding them in kubectl, it's a performance one, they increase the CPU, memory and bandwidth utilisation, solving this at the client side is not enough.

it's unclear from these comments what CPU/memory is being affected here.

Reply:
+1 on doing on server side, we have lot of querying and storing the k8s resources, and a 40% increase with no specific near time use , can be easily avoided by making a client side change. What version was this managedFields introduced in responses from the server side, we might have to plan for this ?

Reply:
If you're worried about storage, just clear client-side. You only need to clear serverside if there's a problem with network bandwidth.

Reply:
client-side already has some changes, see https://github.com/kubernetes/kubernetes/pull/91946 which dropped those from edit, but I keep on hearing this over and over that folks want to see those stripped also when doing get.